lib/spack/spack/build_systems/cmake.py

```diff
@@ -296,6 +296,17 @@ def std_args(pkg, generator=None):
                 define("CMAKE_PREFIX_PATH", spack.build_environment.get_cmake_prefix_path(pkg)),
             ]
         )
+
+        # Set up CMake CUDA architecture
+        if "cuda_arch" in pkg.spec.variants and pkg.spec.satisfies("^cmake@3.18:"):
+            args.append(define("CMAKE_CUDA_ARCHITECTURES", pkg.spec.variants["cuda_arch"].value))
+
+        # Set up CMake HIP architecture
+        if "amdgpu_target" in pkg.spec.variants and pkg.spec.satisfies("^cmake@3.21:"):
+            args.append(
+                define("CMAKE_HIP_ARCHITECTURES", pkg.spec.variants["amdgpu_target"].value)
+            )
```

Comment:
I would extract these snippets into two methods. Something like:
```python
@staticmethod
def cuda_architectures(pkg):
    ...
```
and have packages call the methods explicitly first. 

The rationale is that, as is, this change has a global impact and no opt-out. We should first check whether or not this breaks builds. If we have just a method, we can modify the packages to call it first, and only later call it automatically as part of `stdargs` (if we discover that all packages follow the same convention). Do you agree?